[DebugInfo] Add dataSize to DIBasicType to add DW_AT_bit_size to _BitInt types

OCHyams opened the pull request:
DW_TAG_base_type DIEs are permitted to have both byte_size and bit_size attributes "If the value of an object of the given type does not fully occupy the storage described by a byte size attribute"

* Add DataSizeInBits to DIBasicType (`DIBasicType(... dataSize: n ...)` in IR) .
* Change Clang to add DataSizeInBits to _BitInt type metadata.
* Change LLVM to add DW_AT_bit_size to base_type DIEs that have non-zero DataSizeInBits.

TODO: Do we need to emit DW_AT_data_bit_offset for big endian targets? See discussion on the PR.

Fixes [#61952](https://github.com/llvm/llvm-project/issues/61952)

Co-authored-by: David Stenberg <david.stenberg@ericsson.com>

OCHyams committed:
[DebugInfo] Emit DW_AT_bit_size for _BitSize types and others

DW_TAG_base_type DIEs are permitted to have both byte_size and bit_size
attributes "If the value of an object of the given type does not fully occupy
the storage described by a byte size attribute"

Change Clang to emit the actual bit-size of _BitInt in debug metadata, and
change LLVM to add DW_AT_bit_size to base_type DIEs when the condition above is
true.

OCHyams committed:
add check to base-type-size.ll

OCHyams commented on llvm/lib/CodeGen/AsmPrinter/DwarfCompileUnit.cpp:
```diff
@@ -1795,6 +1795,11 @@ void DwarfCompileUnit::createBaseTypeDIEs() {
     // Round up to smallest number of bytes that contains this number of bits.
     addUInt(Die, dwarf::DW_AT_byte_size, std::nullopt,
             divideCeil(Btr.BitSize, 8));
+    // If the value of an object of the given type does not fully occupy the
```
Good spot, I've added an additional CHECK to that test.

OCHyams commented:
Thanks for taking a look

OCHyams commented:
Breaking out of the inline comments - I've realised this patch doesn't always give us the expected `byte_size`. E.g. For `_BitInt(23)` GCC gives us:

```
0x000000a1:   DW_TAG_base_type
                DW_AT_byte_size (0x04)
                DW_AT_encoding  (DW_ATE_signed)
                DW_AT_bit_size  (0x17)
                DW_AT_name      ("_BitInt(23)")
```

Whereas this patch gives us:
```
0x00000042:   DW_TAG_base_type
                DW_AT_name      ("_BitInt")
                DW_AT_encoding  (DW_ATE_signed)
                DW_AT_byte_size (0x03)
                DW_AT_bit_size  (0x17)
```
Notice Clang says it's 3 bytes (because it's just finding the next byte size that 0x17 bits fits into) whereas GCC says it's 4 bytes (which reflects the `sizeof` size).

It's probably cleanest to pass both byte_size and bit_size down from the front end. I'll have a go at that later/tomorrow if others agree.

OCHyams committed:
Add/use DataSizeInBits (data_size) field to DIBasicType

OCHyams committed:
undo change to base-type-size.ll

OCHyams committed:
Remove DwarfCompileUnit change for DIExpression types

Michael137 commented on llvm/include/llvm/IR/DebugInfoMetadata.h:
```diff
@@ -891,96 +891,114 @@ class DIBasicType : public DIType {
   friend class MDNode;
 
   unsigned Encoding;
+  /// Describes the number of bits used by the value of the object Non-zero when
```
```suggestion
  /// Describes the number of bits used by the value of the object. Non-zero when
```
?

OCHyams committed:
add TODO

OCHyams committed:
Update llvm/include/llvm/IR/DebugInfoMetadata.h

Co-authored-by: Michael Buch <michaelbuch12@gmail.com>

OCHyams commented on llvm/lib/Bitcode/Reader/MetadataLoader.cpp:
```diff
@@ -1540,13 +1540,13 @@ Error MetadataLoader::MetadataLoaderImpl::parseOneMetadata(
                                 : DINode::FlagZero;
     uint32_t NumExtraInhabitants = (Record.size() > 7) ? Record[7] : 0;
-
+    uint32_t DataSizeInBits = (Record.size() > 8) ? Record[8] : 0;
```
If I remove the `(Record.size() > 7)` check for `NumExtraInhabitants ` these tests fail, if that's what you were looking for?
```
Failed Tests (22):
  LLVM :: Bitcode/DIExpression-4.0.ll
  LLVM :: Bitcode/DIExpression-aggresult.ll
  LLVM :: Bitcode/DIExpression-deref.ll
  LLVM :: Bitcode/DIExpression-minus-upgrade.ll
  LLVM :: Bitcode/DIGlobalVariableExpression.ll
  LLVM :: Bitcode/DIGlobalVariableExpression2.ll
  LLVM :: Bitcode/DIImportedEntity_backward.ll
  LLVM :: Bitcode/DIModule-fortran-module.ll
  LLVM :: Bitcode/DINamespace.ll
  LLVM :: Bitcode/DISubprogram-v5.ll
  LLVM :: Bitcode/DITemplateParameter-5.0.ll
  LLVM :: Bitcode/apple-clang-700-compat.test
  LLVM :: Bitcode/diglobalvariable-3.8.ll
  LLVM :: Bitcode/dilocalvariable-3.9.ll
  LLVM :: Bitcode/disubrange-v0.ll
  LLVM :: Bitcode/fortranSubrangeBackward.ll
  LLVM :: Bitcode/invalid.test
  LLVM :: Bitcode/upgrade-cu-locals.ll
  LLVM :: Bitcode/upgrade-dbg-addr.ll
  LLVM :: Bitcode/upgrade-dbg-value.ll
  LLVM :: Bitcode/upgrade-pointer-address-space.ll
  LLVM :: ThinLTO/X86/drop-debug-info.ll
  ```

OCHyams committed:
rm accidental file

OCHyams committed:
data_size -> dataSize

OCHyams committed:
fmt

OCHyams commented:
> I'd be fine with landing this without the endian-ness issue resolved. Perhaps you could add a note about in the commit message, and add TODOs about potentially emitting the `DW_AT_data_bit_offset` attribute for big-endian targets at the two points that the `DW_AT_bit_size` attribute is emitted?

* I've removed the `DwarfCompileUnit` call site as that was only used by types referenced DIExpressions by DW_OP_LLVM_convert (not coming from DIBasicTypes).
* Added a TODO to the `DwarfUnit` mentioning the _data_bit_offset
* Changed the commit message (PR description).

How does this look?

What would you like me to put as your name/email for co-authorship? (this was based on your patch and now looks more or less exactly like it!)

dstenb reviewed (approved):
> How does this look?

LGTM. Thanks for picking up and completing this!

> What would you like me to put as your name/email for co-authorship? (this was based on your patch and now looks more or less exactly like it!)

You can use "David Stenberg \<david.stenberg@ericsson.com\>".

OCHyams committed:
Merge branch 'main' into bitint

OCHyams committed:
data_size -> dataSize in Clang test

OCHyams committed:
data_size -> dataSize in bitcode test
